Low garbage scan data access

robinschmid opened the pull request:
Hey everyone,

I have implemented a class to loop over all scans in a raw data file and access either the raw data or the centroid mass lists. 

- RawDataFile keeps count of the maxNumberOfRawDataPoints / maxNumberOfCentroidDataPoints (over all scans)
- the new class uses a dataFile, scan selection, and RAW/CENTROID flag
- Iterates over all scans and uses only two arrays (mz, intensity)
- gets the mzs and intensities with the getIntensityValues(double[] dst) methods
- It implements the MassSpectrum interface - so no need to change the mass detectors, I only changed the general MassDetectorTask

What do you think?
My idea is that it will promote low garbage + memory efficient access to Scan/MassList data. And we can control it and add some security to it.

SteffenHeu commented on src/main/java/io/github/mzmine/datamodel/impl/MemoryEfficientDataAccess.java:
```diff
@@ -0,0 +1,135 @@
+/**
+ * @author Robin Schmid (https://github.com/robinschmid)
+ */
+public class MemoryEfficientDataAccess {
```
It does not yet implement MassSpectrum, or am i blind?

robinschmid replied:
oh thats weird it somehow did not push. 
I pushed again

SteffenHeu commented on src/main/java/io/github/mzmine/datamodel/listeners/MassListChangedListener.java:
```diff
@@ -0,0 +1,37 @@
+ *
+ * @author Robin Schmid (https://github.com/robinschmid)
+ */
+public interface MassListChangedListener {
```
I don't really like the Idea of having yet another listener for a single task that is only executed in a single module. Seems like unnecessary ram consumption to me. I think having a simple method in the mass detection task or so updating the max number of data points is way more efficient.

robinschmid replied:
No idea we can also just directly call RawDataFile.massListChanged(); and set the maxNumberOfCentroidDataPoints to zero

I thought maybe there are multiple methods that want to listen for changes. Like charts could register and update if things are changed.

SteffenHeu replied:
In which case would that be needed? The mass detector preview does it manuallly anyway.

SteffenHeu commented on src/main/java/io/github/mzmine/project/impl/RawDataFileImpl.java:
```diff
@@ -60,16 +63,21 @@
   private final Hashtable<Integer, Range<Double>> dataMZRange = new Hashtable<>();
-  private final Hashtable<Integer, Range<Float>> dataRTRange = new Hashtable<>();;
+  private final Hashtable<Integer, Range<Float>> dataRTRange = new Hashtable<>();
+  ;
```
Your formatter shows us we have some leftover semicolons here :D

robinschmid replied:
done

SteffenHeu commented on src/main/java/io/github/mzmine/project/impl/RawDataFileImpl.java:
```diff
@@ -263,10 +295,22 @@ public double getDataMaxTotalIonCurrent(int msLevel) {
   @Override
   public synchronized void addScan(Scan newScan) throws IOException {
     scans.add(newScan);
+    if (newScan.getNumberOfDataPoints() > maxRawDataPoints) {
```
This is imo how it should be handled for mass lists during mass detection, too.

robinschmid replied:
We have plenty of modules that create new mass lists and this should be built in somewhere. 
But we can just set the maxNumberOfCentroidDataPoints to -1 each time a mass list is changed in a RawDataFIle.

SteffenHeu replied:
Ok. Just keep in mind that adding listeners to all mobility scans is a huge ram consumption. Convenience is nice, but we have  to keep both sides in mind.

SteffenHeu commented on src/main/java/io/github/mzmine/datamodel/impl/MemoryEfficientDataAccess.java:
```diff
@@ -0,0 +1,135 @@
+
+  protected final RawDataFile dataFile;
+  protected final DataType type;
+  private final Scan[] scans;
```
About this idea in general: If we use it, it should also be able to handle MobilityScans. List<? extends MassSpectrum> here maybe?

robinschmid replied:
I thought about using a factory style and different implementations
EfficientDataAccess.of(RawDataFile, ScanSelection, DataType.CENTROID/RAW)
EfficientDataAccess.of(FeatureList, HandleZero.INCLUDE/EXCLUDE)

SteffenHeu replied:
Ok, so there should be an 
`class MemoryEfficientScanDataAccess implements EffiecientDataAccess`
and
`class MemoryEffiecientMobilityScanDataAccess implements EfficientDataAccess`
and
`class DataAccessFactory`
?
And i can see you also want to use feature lists. Then we should think of an interface and it's methods now so we can handle all those cases and don't have to refactor multiple times after merging this.

robinschmid commented:
I removed the MassListChangedListener and now we call applyMassListChanged(Scan, oldMasses, newMasses) directly on the RawDataFile to reset some precomputed values.
+ cleaned up. 
Ready for merge from my side

SteffenHeu commented on src/main/java/io/github/mzmine/datamodel/data_access/FeatureDataAccess.java:
```diff
@@ -0,0 +1,364 @@
+        case INCLUDE_ZEROS -> {
+          // add detected data points and zero for missing values
+          allScans = flist.getSeletedScans(feature.getRawDataFile());
+          List<Scan> detectedScans = feature.getScanNumbers();
```
This should be featureData.getSpectra(). Is getScanNumbers not deprecated yet?

SteffenHeu commented on src/main/java/io/github/mzmine/datamodel/data_access/FeatureDataAccess.java:
```diff
@@ -0,0 +1,364 @@
+          int detectedIndex = 0;
+          for (int i = 0; i < intensities.length; i++) {
+            if (allScans.get(i) == detectedScans.get(detectedIndex)) {
+              intensities[i] = featureData.getIntensity(detectedIndex);
```
If we want this to be more performant, we should read the whole array/double buffer at once and then loop over the array instead of querieing each intensity.
Before, we (I) should update the feature model to not expose the double buffers to the user, similarly to the raw data model.

robinschmid replied:
Yeah I was trying to keep the garbage low. We could try to have a second final array to do this.

robinschmid replied:
Maybe we just change this in the future?

robinschmid replied:
done, uses batch read now

SteffenHeu commented on src/main/java/io/github/mzmine/datamodel/featuredata/MzSeries.java:
```diff
@@ -33,6 +33,14 @@
+   * @param dst results are reflected in this array
+   * @return All m/z values of detected data points.
+   */
+  double[] getMzValues(double[] dst);
```
Ok, nevermind my earlier comment. Thanks for doing that ;)

SteffenHeu commented on src/main/java/io/github/mzmine/datamodel/impl/SimpleMergedMsMsSpectrum.java:
```diff
@@ -187,9 +186,20 @@ public MassList getMassList() {
     this.massList = massList;
+
+    if (rawDataFile != null) {
+      rawDataFile.applyMassListChanged(this, old, massList);
```
Since a merged spectrum is not part of the original data file, this call should have no effect, or?

robinschmid replied:
No idea. But I think as long as a MergedSpectrum can have a data file - we should call the change method

robinschmid replied:
But I also do not really know where you use the MergedMSMSSpec

SteffenHeu commented on src/main/java/io/github/mzmine/datamodel/data_access/FeatureDataAccess.java:
```diff
@@ -0,0 +1,364 @@
+public class FeatureDataAccess implements IonTimeSeries<Scan> {
+
+  /**
+   * Different types to handle feature data: {@link #ONLY_DETECTED}: Use only detected data points;
```
ONLY_DETECTED should be more specific:
ONLY_DETECTED: only data points currently assigned to the feature. Features might include leading an/or trailing zeros depending on the state of processing. Leading and trailing zeros are added during chromatogram detection, bit might be removed during feature resolving or other processing steps.

robinschmid reviewed (commented):
somehow I cannot comment directly on yours

robinschmid commented:
Should be ready for merging now. @SteffenHeu

SteffenHeu commented:
Looks good :+1:
